crates/rust-analyzer/tests/rust-analyzer/support.rs

```diff
@@ -233,6 +233,70 @@ impl Server {
         });
         self
     }
+
+    pub(crate) fn wait_until_workspace_and_build_data_are_loaded(self) -> Server {
```

Comment:
The logic is quite complicated. But the problem here is our CI is not good on multi threading :( When priming caches, it eats a lot of cpu and coming requests will be too slow to run. 

The second problem here is , the ordering of the "roots_scan" events are based on how threads are run, such that is indeterminate. We tried out best here to guess when the actual "roots scan" ends (And we will have 2 roots scan event here)

Reply:
For cache priming, I think we might want to just disable it for tests? 

For the second issue, I think this logic belongs to the main loop. Specifically (and not only in tests), rust-analyzer should be able to tell the clients: "the project is fully loaded and I can serve every request". 

We generally use the `status` field to track this event:

https://github.com/rust-analyzer/rust-analyzer/blob/ad3cb2125dee4379c11f447216f510d6544c10c5/crates/rust-analyzer/src/main_loop.rs#L381-L395

And we communicate this to client via the status notification:

https://github.com/rust-analyzer/rust-analyzer/blob/ad3cb2125dee4379c11f447216f510d6544c10c5/crates/rust-analyzer/src/reload.rs#L96-L99

I am not sure why we even use progress rathe than status in support, I am trying to fix this right now.

For this PR,I think we want to change the status logic to only become ready once external resources are loaded

Reply:
>I am trying to fix this right now.

Hm, this is for some reason more complicated that I expect. For some reason, multitheraded tests are really slow even with --release...